File: npm/src/__tests__/ApexParserTest.ts
import {
    LiteralContext, Arth1ExpressionContext, CompilationUnitContext,
    QueryContext, StatementContext, TriggerUnitContext
} from "../ApexParser";
import { ThrowingErrorListener, SyntaxException } from "../ThrowingErrorListener";
import { createParser } from "./SyntaxErrorCounter";

test('Boolean Literal', () => {

    const [parser, errorCounter] = createParser("true")
    const context = parser.literal()

    expect(errorCounter.getNumErrors()).toEqual(0)
    expect(context).toBeInstanceOf(LiteralContext)
    expect(context.BooleanLiteral()).toBeTruthy()
    expect(context.BooleanLiteral().text).toBe("true")
})

test('Expression', () => {
    const [parser, errorCounter] = createParser("a * 5")
    const context = parser.expression()

    expect(errorCounter.getNumErrors()).toEqual(0)
    expect(context).toBeInstanceOf(Arth1ExpressionContext)
    const arthExpression = context as Arth1ExpressionContext
    expect(arthExpression.expression().length).toBe(2)
})

test('Compilation Unit', () => {
    const [parser, errorCounter] = createParser("public class Hello {}")

    const context = parser.compilationUnit()

    expect(context).toBeInstanceOf(CompilationUnitContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('Compilation Unit (case insensitive)', () => {
    const [parser, errorCounter] = createParser("Public CLASS Hello {}")

    const context = parser.compilationUnit()

    expect(context).toBeInstanceOf(CompilationUnitContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('Compilation Unit (bug test)', () => {
    const [parser, errorCounter] = createParser(`public class Hello {
        public testMethod void func() {
            System.runAs(u) {
            }
        }
    }`)
    const context = parser.compilationUnit()

    expect(context).toBeInstanceOf(CompilationUnitContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('Compilation Unit (inline SOQL)', () => {
    const [parser, errorCounter] = createParser(`public class Hello {
        public void func() {
            List<Account> accounts = [Select Id from Accounts];
        }
    }`)
    const context = parser.compilationUnit()

    expect(context).toBeInstanceOf(CompilationUnitContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('Compilation Unit (throwing errors)', () => {
    const [parser] = createParser("public class Hello {")

    parser.removeErrorListeners()
    parser.addErrorListener(new ThrowingErrorListener());

    try {
        parser.compilationUnit()
        expect(true).toBe(false)
    } catch (ex) {
        expect(ex).toBeInstanceOf(SyntaxException)
    }
})

test('Trigger Unit', () => {
    const [parser, errorCounter] = createParser("trigger test on Account (before update, after update) {}")
    const context = parser.triggerUnit()

    expect(context).toBeInstanceOf(TriggerUnitContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('SOQL Query', () => {
    const [parser, errorCounter] = createParser("Select Id from Account")
    const context = parser.query()

    expect(context).toBeInstanceOf(QueryContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('SOQL Query Using Field function', () => {
    const [parser, errorCounter] = createParser("Select Fields(All) from Account")

    const context = parser.query()

    expect(context).toBeInstanceOf(QueryContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('CurrencyLiteral', () => {
    const [parser, errorCounter] = createParser("SELECT Id FROM Account WHERE Amount > USD100.01 AND Amount < USD200")

    const context = parser.query()

    expect(context).toBeInstanceOf(QueryContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('IdentifiersThatCouldBeCurrencyLiterals', () => {
    const [parser, errorCounter] = createParser("USD100.name = 'name';")

    const context = parser.statement()

    expect(context).toBeInstanceOf(StatementContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('DateTimeLiteral', () => {
    const [parser, errorCounter] = createParser("SELECT Name, (SELECT Id FROM Account WHERE createdDate > 2020-01-01T12:00:00Z) FROM Opportunity")

    const context = parser.query()

    expect(context).toBeInstanceOf(QueryContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('testNegativeNumericLiteral', () => {
    const [parser, errorCounter] = createParser("SELECT Name FROM Opportunity WHERE Value = -100.123")

    const context = parser.query()

    expect(context).toBeInstanceOf(QueryContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('testLastQuarterKeyword', () => {
    const [parser, errorCounter] = createParser("SELECT Id FROM Account WHERE DueDate = LAST_QUARTER")

    const context = parser.query()

    expect(context).toBeInstanceOf(QueryContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('testSemiAllowedAsWhileBody', () => {
    const [parser, errorCounter] = createParser("while (x++ < 10 && !(y-- < 0));")

    const context = parser.statement()

    expect(context).toBeInstanceOf(StatementContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('testSemiAllowedAsForBody', () => {
    const [parser, errorCounter] = createParser("for(x=0; x<10; x++);")

    const context = parser.statement()

    expect(context).toBeInstanceOf(StatementContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

test('testSemiDisallowedAsGeneralStatement', () => {
    const [parser, errorCounter] = createParser("if (x == 3); else { ; }")

    const context = parser.statement()

    expect(context).toBeInstanceOf(StatementContext)
    expect(errorCounter.getNumErrors()).toEqual(1)
})

test('testWhenLiteralParens', () => {
    const [parser, errorCounter] = createParser(`
    switch on (x) {
        when 1 { return 1; }
        when ((2)) { return 2; }
        when (3), (4) { return 3; }
     }`);

    const context = parser.statement()

    expect(context).toBeInstanceOf(StatementContext)
    expect(errorCounter.getNumErrors()).toEqual(0)
})

File: jvm/src/test/java/com/nawforce/apexparser/ApexParserTest.java
/*
 Copyright (c) 2021 Kevin Jones, All rights reserved.
 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.
 */
package com.nawforce.apexparser;

import org.junit.jupiter.api.Test;

import java.util.Map;

import static com.nawforce.apexparser.SyntaxErrorCounter.createParser;
import static org.junit.jupiter.api.Assertions.*;

public class ApexParserTest {

    @Test
    void testBooleanLiteral() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("true");

        ApexParser.LiteralContext context = parserAndCounter.getKey().literal();
        assertNotNull(context);
        assertEquals("true", context.BooleanLiteral().getText());
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testExpression() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("a * 5");
        ApexParser.ExpressionContext context = parserAndCounter.getKey().expression();
        assertTrue(context instanceof ApexParser.Arth1ExpressionContext);
        assertEquals(2, ((ApexParser.Arth1ExpressionContext) context).expression().size());
    }

    @Test
    void testClass() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("public class Hello {}");
        ApexParser.CompilationUnitContext context = parserAndCounter.getKey().compilationUnit();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testCaseInsensitivity() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("Public CLASS Hello {}");
        ApexParser.CompilationUnitContext context = parserAndCounter.getKey().compilationUnit();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testClassWithError() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("public class Hello {");
        ApexParser.CompilationUnitContext context = parserAndCounter.getKey().compilationUnit();
        assertNotNull(context);
        assertEquals(1, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testClassWithSOQL() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
                "public class Hello {\n" +
                        "        public void func() {\n" +
                        "            List<Account> accounts = [Select Id from Accounts];\n" +
                        "        }\n" +
                        "    }");
        ApexParser.CompilationUnitContext context = parserAndCounter.getKey().compilationUnit();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testTrigger() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser("trigger test on Account (before update, after update) {}");
        ApexParser.TriggerUnitContext context = parserAndCounter.getKey().triggerUnit();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testSemiAllowedAsWhileBody() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
                "while (x++ < 10 && !(y-- < 0));");
        ApexParser.StatementContext context = parserAndCounter.getKey().statement();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testSemiAllowedAsForBody() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
                "for(x=0; x<10; x++);");
        ApexParser.StatementContext context = parserAndCounter.getKey().statement();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testSemiDisallowedAsGeneralStatement() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
                "if (x == 3); else { ; }");
        ApexParser.StatementContext context = parserAndCounter.getKey().statement();
        assertNotNull(context);
        assertEquals(1, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testWhenLiteralParens() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
                "switch on (x) { \n" +
                        "  when 1 { return 1; } \n" +
                        "  when ((2)) { return 2; } \n" +
                        "  when (3), (4) { return 3; } \n" +
                        "}");
        ApexParser.StatementContext context = parserAndCounter.getKey().statement();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testWhenLiteralSigns() {
        Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
                "switch on (x) { \n" +
                        "  when -1 { return 1; } \n" +
                        "  when (+2l) { return 1; } \n" +
                        "  when -+-3 { return 3; } \n" +
                        "}");
        ApexParser.StatementContext context = parserAndCounter.getKey().statement();
        assertNotNull(context);
        assertEquals(0, parserAndCounter.getValue().getNumErrors());
    }

    @Test
    void testSoqlModeKeywords() {
        String [] MODES = new String[] { "USER_MODE", "SYSTEM_MODE" };
        for (String mode : MODES) {
            Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
                    String.format("SELECT Id FROM Account WITH %s", mode));
            ApexParser.QueryContext context = parserAndCounter.getKey().query();
            assertNotNull(context);
            assertEquals(0, parserAndCounter.getValue().getNumErrors());
        }
    }

    @Test
    void testDmlModeKeywords() {
        String [] MODES = new String[] { "USER", "SYSTEM" };
        for (String mode : MODES) {
            Map.Entry<ApexParser, SyntaxErrorCounter> parserAndCounter = createParser(
                    String.format("insert as %s contact;", mode));
            ApexParser.StatementContext context = parserAndCounter.getKey().statement();
            assertNotNull(context);
            assertEquals(0, parserAndCounter.getValue().getNumErrors());
        }
    }
}

